### DISCUSSION TITLE
Offline deployment for pip install version of Doccano

### DISCUSSION THREAD

**Discussion by alewisztann94** [doccano/doccano#1386] [Q&A] [answered]
Feature description 
---------

This feature request is related to doccano/tools/offline_deployment/offline_patcher.py

My question is: Is it possible to use the script offline_patcher.py on the pip install version of Doccano to patch our app so it downloads scripts and serves them statically? 

Environment ->
Two VMS:
VM 1 (CentOS 7.9, sudo privileges and access to internet)
VM 2 (Airgapped CentOS 7.9), 
Doccano version 1.4.0 , python 3.6.8  

Hi I'm currently trying to deploy Doccano onto an airgapped Centos 7.9 VM, and have had some success in the past few weeks but am now reaching a roadblock. I have a somewhat unique environment, where I have my own personal laptop with privileges and internet access and I am trying to deploy Doccano into another environment where I have limited sudo privileges and no internet access. 

I initially started by installing the pip version and was able to successfully have it run on VM 2, however there are css / js files and assets that are missing when I run the application on an airgapped machine. This has been raised before in prior issues but there have been no clear solutions to fix this for the pip install version. 

Here are the steps I took to pip install doccano on my CentOS 7.9 VM:
Get Python 3.6.8, activate a virtual environment, pip install django-utils-six, pip install pysqlite3-binary, pip install doccano
go to {venv directory}/lib/python3.6/site-packages/backend/app/settings.py 
add the code to settings.py:
__import__('pysqlite3')
sys.modules['sqlite3'] = sys.modules.pop('pysqlite3')

After this doccano works fine on VM 2 with the exception of the missing fonts and assets.

I recently found @j-frei 's solution and changed it slightly so it worked in a simulated environment (2 CentOS 7.9 VMs, one airgapped and one with internet access), however unfortunately I don't think I'll be able to run docker and docker-compose without sudo privileges on the actual production VM 2 environment, and I have only successfully run the pip install version of doccano on VM 2 before with my limited privileges. 

I have attempted to try altering some code myself but I am quite confused about how doccano is serving it's html, even with @j-frei 's solution I don't really understand what's happening, I know that the nginx dockerfile is changed and that's what is running the patcher in offline_01_2-patch_and_extract_Docker_images.sh , but I don't understand why we use the args /app/dist /app/dist/offline /offline, or what is really going on when we save the docker images and build locally from the tars. I am learning about nuxt and vue in order to try understand how the front end is served but it's still quite confusing for me. 

Any advice on how I should approach this would be greatly appreciated. Thank you!

**Comment by j-frei** [comment, marked as answer]
Hi Lewis,

Thank you for your interest in my offline deployment approach.
I tried to implement a solution for your PIP deployment.
Can you try it out and give me feedback/report issues?
https://github.com/j-frei/doccano
See `docs/advanced/offline_deployment.md` and `tools/offline_deployment/pip`.

Greetings from Augsburg! :)

Cheers,

Johann

PS: I tried to document the offline_patcher.py arguments inside the file itself. You may want to have a look at it.

**Reply by alewisztann94** [reply]
Hi Johann,

Thanks for writing this script! 

To get it working I used a Cent OS 7.9 VM to execute the script and then copied the 'packages' to my airgapped VM with the same OS then pip installed, after that it worked!

I have access to local mirrors and I'd previously been copying over package requirements manually so I'm not sure if I would be able to install from just the packages provided by 'pip download' on a fresh VM, but some other people on the same environment will use their VMs to try install from the same 'packages' folder so we'll see if that works.

The only differences for my environment is that I needed to download 'setuptools-scm' first, and also I still need to alter the settings.py in order to use pysqlite3 as the sqlite version on my VM is outdated. 
 
I had a look at the patcher script again and I think I get it a bit more, I guess I was more confused about the general way in which Doccano itself serves its front end, because the patcher only goes through the files in 'backend/client/dist' and downloads fonts etc into /static/offline. I'm just wondering how going through the 'index.html' in 'backend/client/dist is enough to fix all of the buttons in the whole app, and am confused about the html in the 'backend/server' and which parts of the overall source code we are using, and how vue/nuxt works in doccano.

But that's all stuff I can work on trying to understand, the main thing is that the fix works and I'm very grateful to you thanks!

Lewis
